### PR TITLE
Trajectory slicing made completely Pythonic #918

### LINKED ISSUE #918: Trajectory slicing should be completely Pythonic
### Expected behaviour

Slicing a trajectory should follow all the rules of slicing a Python list
### Actual behaviour

Some slices raise errors (with appropriate error messages)
### Code to reproduce the behaviour

``` python
import MDAnalysis as mda

u = mda.Universe(top, trj)

In [22]: a = range(10)

In [23]: a[:1000]
Out[23]: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [24]: u.trajectory[:1000]
Out[24]: <DCDReader /home/richard/.local/lib/python2.7/site-packages/MDAnalysisTests-0.14.1.dev0-py2.7.egg/MDAnalysisTests/data/adk_dims.dcd with 98 frames of 3341 atoms>

In [25]: u.trajectory[1000::-1]

IndexError: Frame start/stop outside of the range of the trajectory.

In [26]: a[1000::-1]
Out[26]: [9, 8, 7, 6, 5, 4, 3, 2, 1, 0]

```
### Currently version of MDAnalysis:

(run `python -c "import MDAnalysis as mda; print(mda.__version__)"`)

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
Fixes #918 

Changes made in this Pull Request:
 - Function returns the correct value for 'start' and 'stop' instead of raising an error

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/analysis/base.py [hunk 1]
```diff
@@ -27,16 +27,15 @@
...
 import inspect
 import logging
+
 import numpy as np
 import six
-
 from MDAnalysis import coordinates
 from MDAnalysis.core.groups import AtomGroup
 from MDAnalysis.lib.log import ProgressMeter, _set_verbose
+from six.moves import range, zip
```
Comment (jbarnoud) [inline package/MDAnalysis/analysis/base.py]: Please, keep compatibility imports at the top. It helps keeping track of what behaviours change in the module.

File: package/MDAnalysis/coordinates/base.py [hunk 2]
```diff
@@ -124,28 +124,27 @@
...
-import six
-
+import copy
 import warnings
+import weakref
 
 import numpy as np
-import copy
-import weakref
+import six
```
Comment (jbarnoud) [inline package/MDAnalysis/coordinates/base.py]: Same as above.

File: package/MDAnalysis/coordinates/base.py [hunk 3]
```diff
@@ -1306,22 +1307,23 @@ def check_slice_indices(self, start, stop, step):
 
         if step > 0 and stop < start:
             raise IndexError("Stop frame is lower than start frame")
+        elif step > 0 and stop >= nframes:
```
Comment (jbarnoud) [inline package/MDAnalysis/coordinates/base.py]: The exception above should not be raised; instead it should iterate over no frame.

```python
range(4)[4:2:1]  # []
```

File: package/MDAnalysis/coordinates/base.py [hunk 4]
```diff
@@ -1306,22 +1307,23 @@ def check_slice_indices(self, start, stop, step):
...
+            stop = nframes
         elif step < 0 and start < stop:
             raise IndexError("Start frame is lower than stop frame")
-        if not (0 <= start < nframes) or stop > nframes:
-            raise IndexError(
-                "Frame start/stop outside of the range of the trajectory.")
+        elif step < 0 and start >= nframes:
```
Comment (jbarnoud) [inline package/MDAnalysis/coordinates/base.py]: Same as above. The exception on the line above should not be raised.

File: package/MDAnalysis/core/universe.py [hunk 5]
```diff
@@ -52,28 +52,27 @@
...
-import numpy as np
 import logging
-import copy
+import sys
 import uuid
 
 import MDAnalysis
-import sys
+import numpy as np
+import six
```
Comment (jbarnoud) [inline package/MDAnalysis/core/universe.py]: Compatibility import.

File: package/MDAnalysis/core/universe.py [hunk 6]
```diff
@@ -158,7 +157,7 @@ class Universe(object):
     in_memory
         After reading in the trajectory, transfer it to an in-memory
         representations, which allow for manipulation of coordinates.
-    in_memory_step
+    in_memory_frame_interval
```
Comment (jbarnoud) [inline package/MDAnalysis/core/universe.py]: Your working on an outdated version. Please rebase against the latest version of develop.

**Comment by shobhitagarwal1612** [comment]
Ok thanks

**Comment by richardjgowers** [comment]
https://github.com/MDAnalysis/mdanalysis/blob/develop/testsuite/MDAnalysisTests/coordinates/test_reader_api.py#L135

I think this is where the new tests will have to go.

File: testsuite/MDAnalysisTests/coordinates/test_reader_api.py [hunk 7]
```diff
@@ -131,56 +132,75 @@ def _check_slice(self, start, stop, step):
...
+        assert_equal(self.reference[100:], sl)
 
     def test_slice_IE_3(self):
         def sl():
             return list(self.reader[-100:])
-        assert_raises(IndexError, sl)
+
+        ref = self.reference[-100:]
+        count = 0
+        for i in sl():
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/coordinates/test_reader_api.py]: Why not just `len(sl())` ?

Comment (jbarnoud) [inline testsuite/MDAnalysisTests/coordinates/test_reader_api.py]: Also, the point of having the `sl` sub-function was to test the exception. There is no need for such a complicated system anymore; comparing directly `list(self.reader[-100:])` with `self.reference[-100:]` should do just as well. The same goes for some of the tests above.

File: package/MDAnalysis/core/universe.py [hunk 8]
```diff
@@ -381,17 +380,17 @@ def load_new(self, filename, format=None, in_memory=False, **kwargs):
...
-                                 fname=filename,
-                                 trj_n_atoms=self.trajectory.n_atoms))
+                form=self.trajectory.format,
+                top_n_atoms=len(self.atoms),
+                fname=filename,
+                trj_n_atoms=self.trajectory.n_atoms))
 
         if in_memory:
-            self.transfer_to_memory(step=kwargs.get("in_memory_step", 1))
+            self.transfer_to_memory(kwargs.get("in_memory_frame_interval", 1))
```
Comment (jbarnoud) [inline package/MDAnalysis/core/universe.py]: This should not be. You are reverting to an old version.

File: package/MDAnalysis/core/universe.py [hunk 9]
```diff
@@ -381,17 +380,17 @@ def load_new(self, filename, format=None, in_memory=False, **kwargs):
...
 
         return filename, self.trajectory.format
 
-    def transfer_to_memory(self, start=None, stop=None, step=None, verbose=None, quiet=None):
+    def transfer_to_memory(self, frame_interval=1, verbose=None, quiet=None):
```
Comment (jbarnoud) [inline package/MDAnalysis/core/universe.py]: Same here. This is a reversion and should not be here.

File: package/MDAnalysis/core/universe.py [hunk 10]
```diff
@@ -400,11 +399,7 @@ def transfer_to_memory(self, start=None, stop=None, step=None, verbose=None, qui
 
         Parameters
         ----------
-        start: int, optional
-            start reading from the nth frame.
-        stop: int, optional
-            read upto and excluding the nth frame.
-        step : int, optional
+        frame_interval : int, optional
```
Comment (jbarnoud) [inline package/MDAnalysis/core/universe.py]: Reversion here again.

File: package/MDAnalysis/core/universe.py [hunk 11]
```diff
@@ -423,14 +418,14 @@ def transfer_to_memory(self, start=None, stop=None, step=None, verbose=None, qui
             # trajectory file formats
             try:
                 coordinates = self.trajectory.timeseries(
-                    self.atoms, start=start, stop=stop, step=step, format='fac')
+                    self.atoms, format='fac', step=frame_interval)
```
Comment (jbarnoud) [inline package/MDAnalysis/core/universe.py]: Some more reversion.

File: package/MDAnalysis/core/universe.py [hunk 12]
```diff
@@ -423,14 +418,14 @@ def transfer_to_memory(self, start=None, stop=None, step=None, verbose=None, qui
...
             # if the Timeseries extraction fails,
             # fall back to a slower approach
             except AttributeError:
                 pm = ProgressMeter(self.trajectory.n_frames,
-                                   interval=step, verbose=verbose)
+                                   interval=frame_interval, verbose=verbose)
```
Comment (jbarnoud) [inline package/MDAnalysis/core/universe.py]: Reversion.

File: package/MDAnalysis/core/universe.py [hunk 13]
```diff
@@ -423,14 +418,14 @@ def transfer_to_memory(self, start=None, stop=None, step=None, verbose=None, qui
...
                 coordinates = []  # TODO: use pre-allocated array
-                for ts in self.trajectory[start:stop:step]:
+                for ts in self.trajectory[::frame_interval]:
```
Comment (jbarnoud) [inline package/MDAnalysis/core/universe.py]: Reversion.

**Comment by jbarnoud** [comment]
I did not have the time to do a full review yet, Though, I commented in-line the things that caught my eye.

Also, I guess you have a plug-in in your text editor that reorders the imports. Could you please deactivate it? At least some of the imports are ordered the way for clarity reasons. For instance, the compatibility imports are at the top so we can quickly see what behaviors are not going to be standard, the imports about tests are roughly together. You seem to alphabetize the imports, which is good but only makes sense within a semantic group of them.

File: testsuite/MDAnalysisTests/coordinates/test_reader_api.py [hunk 14]
```diff
@@ -131,56 +132,71 @@ def _check_slice(self, start, stop, step):
...
+            (5, None, -1),
+            (None, 5, -1),
+            (100, 10, 1),
+            (-10, None, 1),
+            (100, None, -1),  # beyond real end
+            (100, 5, -20)
         ]:
             yield self._check_slice, start, stop, step
 
     def test_slice_IE_1a(self):
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/coordinates/test_reader_api.py]: This test should get removed as it no longer raises an Error, and the values for the testing (`[5:1:1]`) should get moved into the test generator above.  The same then applies for the other tests that now work

Comment (shobhitagarwal1612) [inline testsuite/MDAnalysisTests/coordinates/test_reader_api.py]: Ok
:+1:

Comment (shobhitagarwal1612) [inline testsuite/MDAnalysisTests/coordinates/test_reader_api.py]: @richardjgowers Ok

**Comment by shobhitagarwal1612** [comment]
@richardjgowers @jbarnoud I have made all the requested changes. Please review the changes.

File: package/MDAnalysis/analysis/base.py (hunk 1 above)
Comment (richardjgowers) [inline package/MDAnalysis/analysis/base.py]: @shobhitagarwal1612 This still isn't done

**Comment by richardjgowers** [comment]
@shobhitagarwal1612 looks good except the import order.

**Comment by shobhitagarwal1612** [comment]
@richardjgowers Imports reordered...

**Comment by shobhitagarwal1612** [comment]
@richardjgowers Any more changes required?

**Comment by kain88-de** [comment]
@jbarnoud is this OK now?

**Comment by jbarnoud** [comment]
There is a few things I would like to test. But I have the head under the water right now. I'll do my tests as soon as possible.

**Comment by jbarnoud** [comment]
I did some tests. The stop value we get when step is negative and stop is `None` does work as expected. The problem is demonstrated in [this notebook](https://gist.github.com/jbarnoud/0250375eee03ed31bbb0a46c6fbd68d4). I do not know how to solve the issue, though.

**Comment by shobhitagarwal1612** [comment]
@jbarnoud Thanks, I found the bug. I'll make the changes and send the PR

**Comment by shobhitagarwal1612** [comment]
@jbarnoud It is isn't a bug. The function `check_slice_indices` converts all possible `start`, `stop`, `step` for the trajectory slicing as it should do for a normal list. But it is not necessary that the converted values will work similarly for the list as well. Also, the test_cases `(9, None, -1)` and `(None, None, -1)` are already covered in `TestSingleFrameReader`

**Comment by jbarnoud** [comment]
After some digging I found where the issue comes from. There is indeed an issue as the code as it is in this PR work well for slicing trajectories, but breaks analysis tools.

The test checks that we can do this:

```python
[ts.frame for ts in u.trajectory[9:None:-1]]
```

And indeed we can. What we cannot do anymore though, is using `step=-1` in the constructor of the analysis tools. Here is a custom analysis for test purposes:

```python
from MDAnalysis.analysis.base import AnalysisBase

class ReadFrameNumber(AnalysisBase):
    def _prepare(self):
        self.result = []

    def _single_frame(self):
        self.result.append(self._ts.frame)


print(ReadFrameNumber(u.trajectory).run().result)
print(ReadFrameNumber(u.trajectory, step=-1).run().result)
```

The first call to the analysis output the frame number for all the frames as expected. It is the perfect equivalent to `[ts.frame for ts in u.trajectory]`.

The second call should be the equivalent of `[ts.frame for ts in u.trajectory[::-1]`. It should output the same list as before but reversed. Instead it outputs an empty list.

This is because when we slice the trajectory (`u.trajectory[::-1]`), the values returned by `check_slice_indices` are passed to `range`. Instead, in `AnalysisBase` the values returned by `check_slice_indices` are used to splice the trajectory. This creates a discrepancy because these two lines are not equivalent:

```python
range(10, -1, -1)  # [10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
range(10)[10:-1:-1]  # []
```

A solution would be to modify `AnalysisBase` so it does not call `check_slice_indices`, but instead pass directly the `start`, `stop`, and `step` values passed in the constructor. Especially as `check_slice_indices` is not going to raise much exceptions anymore.

@richardjgowers any thoughts?

**Comment by richardjgowers** [comment]
@jbarnoud yeah it looks like `check` isn't working, good catch.  I still think we need some method to know the number of frames we're going to analyse a priori, for things like allocating arrays, progress bars and eventually parallelism.

**Comment by richardjgowers** [comment]
Also, in terms of the issue originally targeted by this PR, can we merge this and open a new issue for `AnalysisBase` with negative skip?

**Comment by jbarnoud** [comment]
My issue with merging is that it means knowingly breaking develop.

**Comment by jbarnoud** [comment]
I'll try to find a workaround in AnalysisBase tomorrow so we can merge this very soon.

**Comment by jbarnoud** [comment]
After some investigation, it seem that we can fix the `AnalysisBase` problem by not saving the modified `start`, `stop`, and `step` returned by `check_slice_indices` as `self.start`, `self.stop`, and `self.step`. We still need to get them to calculate `self.n_frames`. In practice, this means [the call to `check_slice_indices`](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/base.py#L129) should be move after we define `self.start`, `self.stop`, and `self.step`, but before we calculate `self.n_frames`. So it should move right after [line 132](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/base.py#L132).

By doing this, the slice we pass to iterate over the trajectory in the `run` method has directly what was provided by the user; which works. The number of frame is calculated by passing the `start`, `stop`, and `step` returned by `check_slice_indices` to `range`; which works too.

I tried and it does not break any tests.

The only surprise we could have is if somebody wrote an analysis tool using `AnalysisBase`, and use `self.start`, `self.stop`, or `self.step` directly. From what I see, we do not do it in the library.

I expect `analysis.rms.RMSF` to suffer in the corner cases because it reimplement a big chunk of `AnalysisBase`; but the solution here would be to refactor the class.

I also expect that we will forget about the edge case when using `check_slice_indices`; do it should be documented in the docstring of the function, and this discussion should be referenced in that documentation.

**tl;dr** I can merge the pull request once:
* [ ]  [line 129 of `analysis/base.py`](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/base.py#L129) is moved after  [line 132](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/base.py#L132)
* [ ] a warning is added in the docstring of `check_slice_indices` saying that the values that are returned give the expected result when passed in range but have corner cases when passed in a slice (with a reference to this PR discussion).

**Comment by richardjgowers** [comment]
@jbarnoud sounds sensible

**Comment by shobhitagarwal1612** [comment]
@jbarnoud Thanks for explaining the issue. I'll make all the suggested changes.

**Comment by kain88-de** [comment]
I changed the RMSF analysis already in my alignment refactoring PR this doesn't need to be done anymore.

**Comment by jbarnoud** [comment]
@shobhitagarwal1612 Good! I realize I forgot to ask you one last change, though. Could you please add your change in `package/CHANGELOG`, and add your name in `package/AUTHORS` and `testsuite/AUTHORS`?

**Comment by shobhitagarwal1612** [comment]
@jbarnoud I have made the changes

**Comment by kain88-de** [comment]
@jbarnoud merging this branch will result in an ugly history as @orbeckst complained about recently because of the many merge commits with the upstream/develop branch.  Github doesn't remove them on a merge when they have been done with git directly. A squash and merge should be the best option.

@shobhitagarwal1612 In future PR's it would be nice if you rebase against develop instead of merging with it. Normally this is also only required if github complains that a PR can't be merged due to conflicts.

**Comment by jbarnoud** [comment]
@kain88-de I have the "Squash and merge" button ready. I am just waiting for Travis.

**Comment by shobhitagarwal1612** [comment]
@kain88-de I will keep that in mind in the future PR's

**Comment by jbarnoud** [comment]
Congratulations, your first PR made its way to our code base! Thank you for your contribution.
